File: Advanced Algorithms Sem 6/amortized analysis/accounting_dynamicTable.cpp
#include <iostream>
#include <bits/stdc++.h>
#include <vector>

using namespace std;

int main()
{
    vector<int> s;
   
    int arr[]={8,6,7,5,2,6,7,1,2,3,2,2,2,2,2,2,2,2,2,2};
    int cost=0;
    int maxlen=1;
    int bank=0;
    int c=3;
   
    s.push_back(arr[0]);
    cost=1;
    bank=bank+c-cost;
    cout<<"Table size: "<<maxlen<<"     Actual Cost: "<<cost<<"     Bank: "<<bank<<endl;
   
    for(int i=1;i<20;i++)
    {
        if(maxlen>s.size())
        {
            cost=1;
            bank=bank+c-cost;
            s.push_back(arr[i]);
        }
        else
        {
            maxlen=maxlen*2;
            cost=s.size()+1;
            bank=bank+c-cost;
            s.push_back(arr[i]);
        }
       
        cout<<"Table size: "<<maxlen<<"     Actual Cost: "<<cost<<"     Bank: "<<bank<<endl;
    }

    return 0;
}

File: Advanced Algorithms Sem 6 for pracs/amortized analysis/accounting_dynamictable.py
list = [8,6,7,5,2,6,7,1,2,3,2,2,2,2,2,2,2,2,2,2]
table = []
maxlen = 0
c = 3
bank = 0
cost = 0


table.append(list[0])
maxlen = 1
cost = 1
bank += c - cost
print("Table size: ",maxlen ," Actual Cost: ",cost," Bank: ", bank)
for i in range(1,len(list)):
    if(maxlen>len(table)):
        cost = 1
        bank += c - cost
        table.append(list[i])

    else:
        maxlen = maxlen*2
        cost = len(table)+1
        bank += c - cost
        table.append(list[i])

    print("Table size: ",maxlen ," Actual Cost: ",cost," Bank: ", bank)
